src/node_contextify.cc

```diff
@@ -1470,19 +1489,82 @@ void ContextifyContext::ContainsModuleSyntax(
                                                    id_symbol,
                                                    try_catch);
 
-  bool found_error_message_caused_by_module_syntax = false;
+  bool should_retry_as_esm = false;
   if (try_catch.HasCaught() && !try_catch.HasTerminated()) {
     Utf8Value message_value(env->isolate(), try_catch.Message()->Get());
     auto message = message_value.ToStringView();
 
     for (const auto& error_message : esm_syntax_error_messages) {
       if (message.find(error_message) != std::string_view::npos) {
-        found_error_message_caused_by_module_syntax = true;
+        should_retry_as_esm = true;
         break;
       }
     }
+
+    if (!should_retry_as_esm) {
+      for (const auto& error_message : throws_only_in_cjs_error_messages) {
+        if (message.find(error_message) != std::string_view::npos) {
+          // Try parsing again where the user's code is wrapped within an async
+          // function. If the new parse succeeds, then the error was caused by
+          // either a top-level declaration of one of the CommonJS module
+          // variables, or a top-level `await`.
+          TryCatchScope second_parse_try_catch(env);
+          Local<String> wrapped_code =
+              String::Concat(isolate,
+                             String::NewFromUtf8(isolate, "(async function() {")
+                                 .ToLocalChecked(),
+                             code);
+          wrapped_code = String::Concat(
+              isolate,
+              wrapped_code,
+              String::NewFromUtf8(isolate, "})();").ToLocalChecked());
+          ScriptCompiler::Source wrapped_source =
+              GetCommonJSSourceInstance(isolate,
+                                        wrapped_code,
+                                        filename,
+                                        0,
+                                        0,
+                                        host_defined_options,
+                                        nullptr);
+          ContextifyContext::CompileFunctionAndCacheResult(
+              env,
+              context,
+              &wrapped_source,
+              std::move(params),
+              std::vector<Local<Object>>(),
+              options,
+              true,
+              id_symbol,
+              second_parse_try_catch);
```

Comment:
We don't want to cache the async-wrapped function
```suggestion
          code =
              String::Concat(isolate,
                             String::NewFromUtf8(isolate, "(async function() {")
                                 .ToLocalChecked(),
                             code);
          code = String::Concat(
              isolate,
              code,
              String::NewFromUtf8(isolate, "})();").ToLocalChecked());
          ScriptCompiler::Source wrapped_source =
              GetCommonJSSourceInstance(isolate,
                                        code,
                                        filename,
                                        0,
                                        0,
                                        host_defined_options,
                                        nullptr);
          ScriptCompiler::CompileFunction(
              context,
              &wrapped_source,
              params.size(),
              params.data(),
              0,
              nullptr,
              options,
              v8::ScriptCompiler::NoCacheReason::kNoCacheNoReason);
```

Reply:
Added, and added the test mentioned in https://github.com/nodejs/node/pull/52024#discussion_r1519111028